sharedbuffers/mapped_struct.py

```diff
@@ -106,8 +112,256 @@ def _likerobuffer(buf):
+            return self.preloaded[key]
+        orv = self.objmap.pop(key, NONE)
+        rv = self.idmap.pop(key, NONE)
+        self.strong_refs.pop(key, None)
+        if orv is NONE:
+            rv = NONE
+        if rv is not NONE:
+            return rv
+        elif default is NONE:
+            raise KeyError(key)
+        else:
+            return default
+
+    @cython.ccall
+    @cython.returns(cython.bint)
+    def link(self, key, obj):
+        if isinstance(obj, (proxied_list, proxied_dict, BufferProxyObject)):
+            # There's no need to link lifecycles, the object
+            # is identified uniquely by its buffer mapping and that is a stable id
+            return True
+        if is_equality_key(key):
+            # These are equality keys, they are hard references already
+            return True
+        if self.stable_set is not None:
+            if key in self.stable_set:
+                return True
+            elif is_wrapped_key(key) and get_wrapped_key(key) in self.stable_set:
+                return True
+
+        try:
+            self.objmap[key] = obj
+            return True
+        except TypeError:
+            # Not weakly referenceable, try to hold a strong reference then to stop
+            # its id from being reused while we hold its idmap entry
+            self.strong_refs[key] = obj
+            return False
+
+
+@cython.ccall
+@cython.inline
+@cython.returns(cython.bint)
+def is_equality_key(obj):
+    # singletons and small strings
+    if obj is None or obj is () or (isinstance(obj, basestring) and len(obj) < 16):
+        return True
+    elif is_wrapped_key(obj):
+        return is_equality_key(get_wrapped_key(obj))
+    else:
+        return False
+
+
+@cython.ccall
+@cython.locals(lobj = proxied_list, dobj = proxied_dict, oobj = BufferProxyObject)
+def shared_id(obj):
+    # (id(buf), offs) of buffer-mapped proxies
+    if isinstance(obj, proxied_list):
+        lobj = obj
+        return (id(lobj.buf) << 64) | lobj.offs
+    elif isinstance(obj, proxied_dict):
+        dobj = obj
+        return (id(dobj.buf) << 64) | dobj.offs
+    elif isinstance(obj, BufferProxyObject):
+        oobj = obj
+        return (id(oobj.buf) << 64) | oobj.offs
+    elif is_equality_key(obj):
+        # Try an equality key, use the object itself as key if hashable
+        try:
+            hash(obj)
+        except TypeError:
+            pass
+        else:
+            return obj
+
+    # Otherwise plainly the id of the object
+    return id(obj)
+
+
+class WRAPPED:
+    pass
+
+
+@cython.ccall
+def wrapped_id(obj):
+    xid = shared_id(obj)
+    if isinstance(xid, (int, long)):
+        # We keep numeric ids numeric
+        xid |= 0xFL << 64
+    else:
+        # Equality ids are wrapped in a tuple
+        xid = (WRAPPED, xid)
+    return xid
+
+
+@cython.ccall
+@cython.inline
+@cython.returns(cython.bint)
+def is_wrapped_key(obj):
+    # singletons and small strings
```

Comment:
This doesn't seem to handle small strings (Although it should, I think).

Reply:
I think that's a copypasto